autobackup and reboot

```
---- 1.1.10 --------

/opt/meshcentral/node_modules/axios/lib/utils.js:387
        destObj[prop] = sourceObj[prop];
                      ^

TypeError: Cannot assign to read only property 'isAxiosError' of object 'Error: socket hang up'
    at Object.toFlatObject (/opt/meshcentral/node_modules/axios/lib/utils.js:387:23)
    at AxiosError.from (/opt/meshcentral/node_modules/axios/lib/core/AxiosError.js:73:9)
    at PassThrough.handleStreamError (/opt/meshcentral/node_modules/axios/lib/adapters/http.js:418:27)
    at PassThrough.emit (node:events:514:28)
    at /opt/meshcentral/node_modules/webdav/dist/node/operations/createStream.js:88:21
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)

Node.js v18.17.1
```
I enter a command in the autobackup console.

```
 autobackup
Starting auto-backup...
Auto-backup completed.
Attempting WebDAV upload...
Uploading using WebDAV...
WebDAV upload completed
```
the file could not be uploaded to the cloud after 10 minutes (+-) will give an error message and reboot, and after the reboot the file will be uploaded to the cloud.

I use the docker container of the latest version.

my config

```

> {
>   "$schema": "http://info.meshcentral.com/downloads/meshcentral-config-schema.json",
>   "settings": {
>     "cert": "123",
>     "_WANonly": true,
>     "_LANonly": true,
>     "sessionKey": "123",
>     "port": 443,
>     "_aliasPort": 443,
>     "redirPort": 80,
>     "_redirAliasPort": 80,
>     "AgentPong": 300,
>     "TLSOffload": false,
>     "SelfUpdate": false,
>     "AllowFraming": "false",
>     "WebRTC": "false",
>     "AutoBackup": {
>       "backupPath": "/opt/meshcentral/meshcentral-backups",
>       "backupInvervalHours": "24",
>       "keepLastDaysBackup": "10",
>       "zippassword": "123",
>   		"webDAV" : {
>                 "url": "https://webdav.yandex.ru/soft/backup",
>                 "username": "123",
>                 "password": "123",
>                 "folderName": "MeshCentral-Backups",
>                 "maxFiles": "10"
> 	    }
>     }
>   },
>   "domains": {
>         "": {
>         "_title": "MyServer",
>     "_title2": "Servername",
>     "_minify": true,
>     "NewAccounts": "false",
>         "_userNameIsEmail": true,
>     "certUrl": "https://123:443"
>         }
>   },
>   "_letsencrypt": {
>     "__comment__": "Requires NodeJS 8.x or better, Go to https://letsdebug.net/ first before>",
>     "_email": "myemail@mydomain.com",
>     "_names": "myserver.mydomain.com",
>         "production": false
>   }
> }

```

> thanks, i see the same thing! also spotted another bug, `webdav` only gets installed IF you set a `zippassword` but maybe you want to live dangerously and not enter a `zippassword`


I commented out the parameter and restarted docker. After send the command, the program immediately went into reboot
```
2023-09-06T10:53:42.187551438Z Installing archiver@5.3.1...
2023-09-06T10:53:57.773117203Z Installing otplib@10.2.3...
2023-09-06T10:54:10.524776436Z Installing archiver@5.3.1...
2023-09-06T10:54:22.539962773Z MeshCentral HTTP redirection server running on port 80.
2023-09-06T10:54:22.672585852Z MeshCentral v1.1.10, Hybrid (LAN + WAN) mode, Production mode.
2023-09-06T10:54:28.772325453Z MeshCentral Intel(R) AMT server running on help.123.ru:4433.
2023-09-06T10:54:29.027647686Z MeshCentral HTTPS server running on help.123.ru:443.
2023-09-06T10:54:29.066321685Z Loaded web certificate from "https://npm.123.ru:443", host: "help.123.ru"
2023-09-06T10:54:29.068075534Z   SHA384 cert hash: f9cf9560184c5e6b53f26e61b7194d192bc01dee3a2b6418cd4bd211fcd1d1baaba032944a99311812979bd3fed3e001
2023-09-06T10:54:58.124430764Z ERR: node:internal/modules/cjs/loader:1080
2023-09-06T10:54:58.124588954Z   throw err;
2023-09-06T10:54:58.124647411Z   ^
2023-09-06T10:54:58.124695113Z 
2023-09-06T10:54:58.124741138Z Error: Cannot find module 'webdav'
2023-09-06T10:54:58.124788211Z Require stack:
2023-09-06T10:54:58.124836401Z - /opt/meshcentral/node_modules/meshcentral/db.js
2023-09-06T10:54:58.124884522Z - /opt/meshcentral/node_modules/meshcentral/meshcentral.js
2023-09-06T10:54:58.124933271Z     at Module._resolveFilename (node:internal/modules/cjs/loader:1077:15)
2023-09-06T10:54:58.124983626Z     at Module._load (node:internal/modules/cjs/loader:922:27)
2023-09-06T10:54:58.125036077Z     at Module.require (node:internal/modules/cjs/loader:1143:19)
2023-09-06T10:54:58.125087550Z     at require (node:internal/modules/cjs/helpers:121:18)
2023-09-06T10:54:58.125138394Z     at obj.performCloudBackup (/opt/meshcentral/node_modules/meshcentral/db.js:3075:38)
2023-09-06T10:54:58.125187283Z     at WriteStream.<anonymous> (/opt/meshcentral/node_modules/meshcentral/db.js:2997:130)
2023-09-06T10:54:58.125234914Z     at WriteStream.emit (node:events:514:28)
2023-09-06T10:54:58.125280800Z     at emitCloseNT (node:internal/streams/destroy:132:10)
2023-09-06T10:54:58.125326546Z     at process.processTicksAndRejections (node:internal/process/task_queues:81:21) {
2023-09-06T10:54:58.125370895Z   code: 'MODULE_NOT_FOUND',
2023-09-06T10:54:58.125416012Z   requireStack: [
2023-09-06T10:54:58.125459314Z     '/opt/meshcentral/node_modules/meshcentral/db.js',
2023-09-06T10:54:58.125502266Z     '/opt/meshcentral/node_modules/meshcentral/meshcentral.js'
2023-09-06T10:54:58.125546336Z   ]
2023-09-06T10:54:58.125593129Z }
2023-09-06T10:54:58.125639574Z 
2023-09-06T10:54:58.125687834Z Node.js v18.17.1
2023-09-06T10:54:58.257430777Z Error: Command failed: /usr/local/bin/node /opt/meshcentral/node_modules/meshcentral --launch 8
2023-09-06T10:54:58.257776840Z node:internal/modules/cjs/loader:1080
2023-09-06T10:54:58.257845633Z   throw err;
2023-09-06T10:54:58.257893684Z   ^
2023-09-06T10:54:58.257940827Z 
2023-09-06T10:54:58.257987550Z Error: Cannot find module 'webdav'
2023-09-06T10:54:58.258033715Z Require stack:
2023-09-06T10:54:58.258168089Z - /opt/meshcentral/node_modules/meshcentral/db.js
2023-09-06T10:54:58.258205175Z - /opt/meshcentral/node_modules/meshcentral/meshcentral.js
2023-09-06T10:54:58.258239048Z     at Module._resolveFilename (node:internal/modules/cjs/loader:1077:15)
2023-09-06T10:54:58.258273829Z     at Module._load (node:internal/modules/cjs/loader:922:27)
2023-09-06T10:54:58.258307841Z     at Module.require (node:internal/modules/cjs/loader:1143:19)
2023-09-06T10:54:58.258341016Z     at require (node:internal/modules/cjs/helpers:121:18)
2023-09-06T10:54:58.258375168Z     at obj.performCloudBackup (/opt/meshcentral/node_modules/meshcentral/db.js:3075:38)
2023-09-06T10:54:58.258410159Z     at WriteStream.<anonymous> (/opt/meshcentral/node_modules/meshcentral/db.js:2997:130)
2023-09-06T10:54:58.258445079Z     at WriteStream.emit (node:events:514:28)
2023-09-06T10:54:58.258478393Z     at emitCloseNT (node:internal/streams/destroy:132:10)
2023-09-06T10:54:58.258513174Z     at process.processTicksAndRejections (node:internal/process/task_queues:81:21) {
2023-09-06T10:54:58.258547047Z   code: 'MODULE_NOT_FOUND',
2023-09-06T10:54:58.258580082Z   requireStack: [
2023-09-06T10:54:58.258612907Z     '/opt/meshcentral/node_modules/meshcentral/db.js',
2023-09-06T10:54:58.258646291Z     '/opt/meshcentral/node_modules/meshcentral/meshcentral.js'
2023-09-06T10:54:58.258679536Z   ]
2023-09-06T10:54:58.258712710Z }
2023-09-06T10:54:58.258745186Z 
2023-09-06T10:54:58.258777453Z Node.js v18.17.1
2023-09-06T10:54:58.258810557Z 
2023-09-06T10:54:58.258843243Z     at ChildProcess.exithandler (node:child_process:419:12)
2023-09-06T10:54:58.258876348Z     at ChildProcess.emit (node:events:526:35)
2023-09-06T10:54:58.258910011Z     at maybeClose (node:internal/child_process:1091:16)
2023-09-06T10:54:58.258943395Z     at Socket.<anonymous> (node:internal/child_process:449:11)
2023-09-06T10:54:58.258977059Z     at Socket.emit (node:events:514:28)
2023-09-06T10:54:58.259010582Z     at Pipe.<anonymous> (node:net:323:12) {
2023-09-06T10:54:58.259044036Z   code: 1,
2023-09-06T10:54:58.259075814Z   killed: false,
2023-09-06T10:54:58.259108918Z   signal: null,
2023-09-06T10:54:58.259141674Z   cmd: '/usr/local/bin/node /opt/meshcentral/node_modules/meshcentral --launch 8'
2023-09-06T10:54:58.259174569Z }
2023-09-06T10:54:58.259254956Z ERROR: MeshCentral failed with critical error, check mesherrors.txt. Restarting in 5 seconds...
2023-09-06T10:55:07.220417780Z MeshCentral HTTP redirection server running on port 80.
2023-09-06T10:55:07.347625386Z MeshCentral v1.1.10, Hybrid (LAN + WAN) mode, Production mode.
2023-09-06T10:55:13.402434508Z MeshCentral Intel(R) AMT server running on help.123.ru:4433.
2023-09-06T10:55:13.603036410Z MeshCentral HTTPS server running on help.123.ru:443.
2023-09-06T10:55:13.642801258Z Loaded web certificate from "https://npm.123.ru:443", host: "help.123.ru"
2023-09-06T10:55:13.644062867Z   SHA384 cert hash: 123

```

> > thanks, i see the same thing! also spotted another bug, `webdav` only gets installed IF you set a `zippassword` but maybe you want to live dangerously and not enter a `zippassword`
> 
> I commented out the parameter and restarted docker. After send the command, the program immediately went into reboot
> 
> ```
> 2023-09-06T10:53:42.187551438Z Installing archiver@5.3.1...
> 2023-09-06T10:53:57.773117203Z Installing otplib@10.2.3...
> 2023-09-06T10:54:10.524776436Z Installing archiver@5.3.1...
> 2023-09-06T10:54:22.539962773Z MeshCentral HTTP redirection server running on port 80.
> 2023-09-06T10:54:22.672585852Z MeshCentral v1.1.10, Hybrid (LAN + WAN) mode, Production mode.
> 2023-09-06T10:54:28.772325453Z MeshCentral Intel(R) AMT server running on help.123.ru:4433.
> 2023-09-06T10:54:29.027647686Z MeshCentral HTTPS server running on help.123.ru:443.
> 2023-09-06T10:54:29.066321685Z Loaded web certificate from "https://npm.123.ru:443", host: "help.123.ru"
> 2023-09-06T10:54:29.068075534Z   SHA384 cert hash: f9cf9560184c5e6b53f26e61b7194d192bc01dee3a2b6418cd4bd211fcd1d1baaba032944a99311812979bd3fed3e001
> 2023-09-06T10:54:58.124430764Z ERR: node:internal/modules/cjs/loader:1080
> 2023-09-06T10:54:58.124588954Z   throw err;
> 2023-09-06T10:54:58.124647411Z   ^
> 2023-09-06T10:54:58.124695113Z 
> 2023-09-06T10:54:58.124741138Z Error: Cannot find module 'webdav'
> 2023-09-06T10:54:58.124788211Z Require stack:
> 2023-09-06T10:54:58.124836401Z - /opt/meshcentral/node_modules/meshcentral/db.js
> 2023-09-06T10:54:58.124884522Z - /opt/meshcentral/node_modules/meshcentral/meshcentral.js
> 2023-09-06T10:54:58.124933271Z     at Module._resolveFilename (node:internal/modules/cjs/loader:1077:15)
> 2023-09-06T10:54:58.124983626Z     at Module._load (node:internal/modules/cjs/loader:922:27)
> 2023-09-06T10:54:58.125036077Z     at Module.require (node:internal/modules/cjs/loader:1143:19)
> 2023-09-06T10:54:58.125087550Z     at require (node:internal/modules/cjs/helpers:121:18)
> 2023-09-06T10:54:58.125138394Z     at obj.performCloudBackup (/opt/meshcentral/node_modules/meshcentral/db.js:3075:38)
> 2023-09-06T10:54:58.125187283Z     at WriteStream.<anonymous> (/opt/meshcentral/node_modules/meshcentral/db.js:2997:130)
> 2023-09-06T10:54:58.125234914Z     at WriteStream.emit (node:events:514:28)
> 2023-09-06T10:54:58.125280800Z     at emitCloseNT (node:internal/streams/destroy:132:10)
> 2023-09-06T10:54:58.125326546Z     at process.processTicksAndRejections (node:internal/process/task_queues:81:21) {
> 2023-09-06T10:54:58.125370895Z   code: 'MODULE_NOT_FOUND',
> 2023-09-06T10:54:58.125416012Z   requireStack: [
> 2023-09-06T10:54:58.125459314Z     '/opt/meshcentral/node_modules/meshcentral/db.js',
> 2023-09-06T10:54:58.125502266Z     '/opt/meshcentral/node_modules/meshcentral/meshcentral.js'
> 2023-09-06T10:54:58.125546336Z   ]
> 2023-09-06T10:54:58.125593129Z }
> 2023-09-06T10:54:58.125639574Z 
> 2023-09-06T10:54:58.125687834Z Node.js v18.17.1
> 2023-09-06T10:54:58.257430777Z Error: Command failed: /usr/local/bin/node /opt/meshcentral/node_modules/meshcentral --launch 8
> 2023-09-06T10:54:58.257776840Z node:internal/modules/cjs/loader:1080
> 2023-09-06T10:54:58.257845633Z   throw err;
> 2023-09-06T10:54:58.257893684Z   ^
> 2023-09-06T10:54:58.257940827Z 
> 2023-09-06T10:54:58.257987550Z Error: Cannot find module 'webdav'
> 2023-09-06T10:54:58.258033715Z Require stack:
> 2023-09-06T10:54:58.258168089Z - /opt/meshcentral/node_modules/meshcentral/db.js
> 2023-09-06T10:54:58.258205175Z - /opt/meshcentral/node_modules/meshcentral/meshcentral.js
> 2023-09-06T10:54:58.258239048Z     at Module._resolveFilename (node:internal/modules/cjs/loader:1077:15)
> 2023-09-06T10:54:58.258273829Z     at Module._load (node:internal/modules/cjs/loader:922:27)
> 2023-09-06T10:54:58.258307841Z     at Module.require (node:internal/modules/cjs/loader:1143:19)
> 2023-09-06T10:54:58.258341016Z     at require (node:internal/modules/cjs/helpers:121:18)
> 2023-09-06T10:54:58.258375168Z     at obj.performCloudBackup (/opt/meshcentral/node_modules/meshcentral/db.js:3075:38)
> 2023-09-06T10:54:58.258410159Z     at WriteStream.<anonymous> (/opt/meshcentral/node_modules/meshcentral/db.js:2997:130)
> 2023-09-06T10:54:58.258445079Z     at WriteStream.emit (node:events:514:28)
> 2023-09-06T10:54:58.258478393Z     at emitCloseNT (node:internal/streams/destroy:132:10)
> 2023-09-06T10:54:58.258513174Z     at process.processTicksAndRejections (node:internal/process/task_queues:81:21) {
> 2023-09-06T10:54:58.258547047Z   code: 'MODULE_NOT_FOUND',
> 2023-09-06T10:54:58.258580082Z   requireStack: [
> 2023-09-06T10:54:58.258612907Z     '/opt/meshcentral/node_modules/meshcentral/db.js',
> 2023-09-06T10:54:58.258646291Z     '/opt/meshcentral/node_modules/meshcentral/meshcentral.js'
> 2023-09-06T10:54:58.258679536Z   ]
> 2023-09-06T10:54:58.258712710Z }
> 2023-09-06T10:54:58.258745186Z 
> 2023-09-06T10:54:58.258777453Z Node.js v18.17.1
> 2023-09-06T10:54:58.258810557Z 
> 2023-09-06T10:54:58.258843243Z     at ChildProcess.exithandler (node:child_process:419:12)
> 2023-09-06T10:54:58.258876348Z     at ChildProcess.emit (node:events:526:35)
> 2023-09-06T10:54:58.258910011Z     at maybeClose (node:internal/child_process:1091:16)
> 2023-09-06T10:54:58.258943395Z     at Socket.<anonymous> (node:internal/child_process:449:11)
> 2023-09-06T10:54:58.258977059Z     at Socket.emit (node:events:514:28)
> 2023-09-06T10:54:58.259010582Z     at Pipe.<anonymous> (node:net:323:12) {
> 2023-09-06T10:54:58.259044036Z   code: 1,
> 2023-09-06T10:54:58.259075814Z   killed: false,
> 2023-09-06T10:54:58.259108918Z   signal: null,
> 2023-09-06T10:54:58.259141674Z   cmd: '/usr/local/bin/node /opt/meshcentral/node_modules/meshcentral --launch 8'
> 2023-09-06T10:54:58.259174569Z }
> 2023-09-06T10:54:58.259254956Z ERROR: MeshCentral failed with critical error, check mesherrors.txt. Restarting in 5 seconds...
> 2023-09-06T10:55:07.220417780Z MeshCentral HTTP redirection server running on port 80.
> 2023-09-06T10:55:07.347625386Z MeshCentral v1.1.10, Hybrid (LAN + WAN) mode, Production mode.
> 2023-09-06T10:55:13.402434508Z MeshCentral Intel(R) AMT server running on help.123.ru:4433.
> 2023-09-06T10:55:13.603036410Z MeshCentral HTTPS server running on help.123.ru:443.
> 2023-09-06T10:55:13.642801258Z Loaded web certificate from "https://npm.123.ru:443", host: "help.123.ru"
> 2023-09-06T10:55:13.644062867Z   SHA384 cert hash: 123
> ```

yep its a bug! because docker images dont perstist, every restart it has to install the missing packages you need, 
so if you need webdav, you need to set a zippassword, but again zippassword should be optional!
ive also spotted another bug where its not creating the folder correctly with my nextcloud webdav
still investigating for you 👍

--- a/db.js
+++ b/db.js
@@ -3064,39 +3064,37 @@ module.exports.CreateDB = function (parent, func) {
 
             // Upload to the WebDAV folder
             function performWebDavUpload(client, filepath) {
-                var fileStream = require('fs').createReadStream(filepath);
-                fileStream.on('close', function () { if (func) { func('WebDAV upload completed'); } })
-                fileStream.on('error', function (err) { if (func) { func('WebDAV (fileUpload) error: ' + err); } })
-                fileStream.pipe(client.createWriteStream('/' + webdavfolderName + '/' + require('path').basename(filepath)));
-                if (func) { func('Uploading using WebDAV...'); }
+                require('fs').stat(filepath, function(err,stat){
+                    var fileStream = require('fs').createReadStream(filepath);
+                    fileStream.on('close', function () { if (func) { func('WebDAV upload completed'); } })
+                    fileStream.on('error', function (err) { if (func) { func('WebDAV (fileUpload) error: ' + err); } })
+                    fileStream.pipe(client.createWriteStream('/' + webdavfolderName + '/' + require('path').basename(filepath), { headers: { "Content-Length": stat.size } }));
+                    if (func) { func('Uploading using WebDAV...'); }
+                });
             }
 
             if (func) { func('Attempting WebDAV upload...'); }
             const { createClient } = require('webdav');
-            const client = createClient(parent.config.settings.autobackup.webdav.url, { username: parent.config.settings.autobackup.webdav.username, password: parent.config.settings.autobackup.webdav.password });
-            var directoryItems = client.getDirectoryContents('/');
-            directoryItems.then(
-                function (files) {
-                    var folderFound = false;
-                    for (var i in files) { if ((files[i].basename == webdavfolderName) && (files[i].type == 'directory')) { folderFound = true; } }
-                    if (folderFound == false) {
-                        client.createDirectory(webdavfolderName).then(function (a) {
-                            if (a.statusText == 'Created') {
-                                if (func) { func('WebDAV folder created'); }
-                                performWebDavUpload(client, filename);
-                            } else {
-                                if (func) { func('WebDAV (createDirectory) status: ' + a.statusText); }
-                            }
-                        }).catch(function (err) {
-                            if (func) { func('WebDAV (createDirectory) error: ' + err); }
-                        });
-                    } else {
-                        performWebDavCleanup(client);
+            const client = createClient(parent.config.settings.autobackup.webdav.url, {
+                username: parent.config.settings.autobackup.webdav.username,
+                password: parent.config.settings.autobackup.webdav.password,
+                maxContentLength: Infinity,
+                maxBodyLength: Infinity
+            });
+            client.exists(webdavfolderName).then(function(a){
+                if(a){
+                    performWebDavCleanup(client);
+                    performWebDavUpload(client, filename);
+                }else{
+                    client.createDirectory(webdavfolderName, {recursive: true}).then(function (a) {
+                        if (func) { func('WebDAV folder created'); }
                         performWebDavUpload(client, filename);
-                    }
+                    }).catch(function (err) {
+                        if (func) { func('WebDAV (createDirectory) error: ' + err); }
+                    });
                 }
-            ).catch(function (err) {
-                if (func) { func('WebDAV (getDirectoryContents) error: ' + err); }
+            }).catch(function (err) {
+                if (func) { func('WebDAV (exists) error: ' + err); }
             });
         }
 

--- a/meshcentral.js
+++ b/meshcentral.js
@@ -4022,12 +4022,12 @@ function mainStart() {
         // Setup encrypted zip support if needed
         if (config.settings.autobackup && config.settings.autobackup.zippassword) {
             modules.push('archiver-zip-encrypted');
-            // Enable Google Drive Support
-            if (typeof config.settings.autobackup.googledrive == 'object') { modules.push('googleapis'); }
-            // Enable WebDAV Support
-            if (typeof config.settings.autobackup.webdav == 'object') {
-                if ((typeof config.settings.autobackup.webdav.url != 'string') || (typeof config.settings.autobackup.webdav.username != 'string') || (typeof config.settings.autobackup.webdav.password != 'string')) { addServerWarning("Missing WebDAV parameters.", 2, null, !args.launch); } else { modules.push('webdav@4.11.2'); }
-            }
+        }
+        // Enable Google Drive Support
+        if (typeof config.settings.autobackup.googledrive == 'object') { modules.push('googleapis'); }
+        // Enable WebDAV Support
+        if (typeof config.settings.autobackup.webdav == 'object') {
+            if ((typeof config.settings.autobackup.webdav.url != 'string') || (typeof config.settings.autobackup.webdav.username != 'string') || (typeof config.settings.autobackup.webdav.password != 'string')) { addServerWarning("Missing WebDAV parameters.", 2, null, !args.launch); } else { modules.push('webdav@4.11.2'); }
         }
 
         // Setup common password blocking

--- a/meshcentral.js
+++ b/meshcentral.js
@@ -4027,7 +4027,7 @@ function mainStart() {
         if (typeof config.settings.autobackup.googledrive == 'object') { modules.push('googleapis'); }
         // Enable WebDAV Support
         if (typeof config.settings.autobackup.webdav == 'object') {
-            if ((typeof config.settings.autobackup.webdav.url != 'string') || (typeof config.settings.autobackup.webdav.username != 'string') || (typeof config.settings.autobackup.webdav.password != 'string')) { addServerWarning("Missing WebDAV parameters.", 2, null, !args.launch); } else { modules.push('webdav@4.11.2'); }
+            if ((typeof config.settings.autobackup.webdav.url != 'string') || (typeof config.settings.autobackup.webdav.username != 'string') || (typeof config.settings.autobackup.webdav.password != 'string')) { addServerWarning("Missing WebDAV parameters.", 2, null, !args.launch); } else { modules.push('webdav@4.11.3'); }
         }
 
         // Setup common password blocking

